### PR TITLE
[#2]  Add CI pipeline for test

### LINKED ISSUE #2: [Chore] Set up CI pipeline for the project
## Why
To build the application we would need require few dependency such as nextJs, Typescript and few testing frameworks to facilitate the development process

## Acceptance Criteria

- Add the required dependencies with nextJS and testing frameworks

## Implementation Details
- Add CI pipeline for validating the linting with eslint and prettier
- Add CI pipeline to validate the build of the project
- Add CI pipeline to validate docker image build
- Add CI pipeline for test

### PR DESCRIPTION
**Pull request by Prabeshpd**
- Close #2 

## What happened 👀

Created a CI pipeline leveraging GitHub actions to ensure build, lint and test

- Created a pipeline that builds the application
- Created a pipeline that lints the application
- Created a pipeline that test the application

## Insight 📝
To ensure there is no breakage on the code merge on every pull request and on merge a pipeline is built that validates whether the build, linting and test of the application runs without error. 

## Proof Of Work 📹

Screen Shot 2566-07-21 at 09 09 41

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -0,0 +1,70 @@
+name: Test
+
+on:
+  pull_request:
+  push:
+    branches:
+      - development
```
Comment (olivierobert) [inline .github/workflows/test.yml:7]: The branch name should be `develop`. Refer to: https://nimblehq.co/compass/development/version-control/branch-management/#default-branches

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+  lint:
+    name: Check Linting
+    runs-on: ubuntu-latest
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v3
+
+      - name: Setup node and restore cached dependencies
```
Comment (olivierobert) [inline .github/workflows/test.yml:42]: SYntax issue:

```suggestion
      - name: Set up node and restore cached dependencies
```

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        uses: actions/setup-node@v3
+        with:
+          node-version: '16'
```
Comment (olivierobert) [inline .github/workflows/test.yml:45]: The current latest LTS version is 18 so we should use that version

**Review by olivierobert — CHANGES_REQUESTED** [review]
There is currently no benefit in splitting the workfklows into 3 jobs 🤔 There is a lot of repetitions and there are no parallelism benefits (linting and test runns will be fast). Only one job might be sufficient. Then later on it might make sense to split UI and Non-UI tests if the test suite becomes too long.

**Comment by Prabeshpd** [comment]
> There is currently no benefit in splitting the workfklows into 3 jobs 🤔 There is a lot of repetitions and there are no parallelism benefits (linting and test runns will be fast). Only one job might be sufficient. Then later on it might make sense to split UI and Non-UI tests if the test suite becomes too long.

Ok. Noted

File: .github/workflows/test.yml (hunk 3 above)
Comment (Prabeshpd) [reply .github/workflows/test.yml:45]: For this one. I tried setting up the project from node 18. Currently nimble eslint package `@nimblehq/eslint-config-nimble"` does not support the node 18. When I looked in the repo there was already a PR. Waiting for that PR to be merged.

Comment (Prabeshpd) [reply .github/workflows/test.yml]: Updated the node version to 18 successfully

File: .github/workflows/test.yml [hunk 4]
```diff
@@ -0,0 +1,38 @@
+name: Test
+
+on:
+  pull_request:
+  push:
+    branches:
+      - develop
+      - main
```
Comment (olivierobert) [inline .github/workflows/test.yml:8]: There is no need to retest on merges. Sinc you work alone, make sure you rebase all your branches before merging them. Running the tests again will not catch more issues.

**Review by olivierobert — CHANGES_REQUESTED** [review]

**Review by tednguyendev — APPROVED** [review]

**Review by olivierobert — APPROVED** [review]

**Pull request merged**